lib/web_ui/lib/src/engine/frame_reference.dart

```diff
@@ -2,28 +2,61 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-/// A monotonically increasing frame number being rendered.
+/// Holds a resource that's valid until the end of the current frame.
 ///
-/// Used for debugging only.
-int debugFrameNumber = 1;
+/// If a frame-scoped object is created outside a frame, it is disposed of at
+/// the end of the next frame.
+///
+/// At the end of the frame the [disposeFrameResources] method is called automatically.
+abstract class FrameScoped {
+  FrameScoped() {
+    _scopedObjects.add(this);
+  }
+
+  /// Disposes of the frame resources held by this object.
+  void disposeFrameResources();
+}
 
-List<FrameReference<dynamic>> frameReferences = <FrameReference<dynamic>>[];
+/// All frame-scoped variables for the current frame.
+List<FrameScoped> _scopedObjects = <FrameScoped>[];
```

Comment:
I think it would be nice if this was a bit more generic. In some cases, especially with asynchronous rendering, we might need to actually have multiple frame scopes alive at the same time. So we should try to make this machinery as instance-based as possible, and then if the rendering strategy really just requires a single scope to exist at a time (like the CanvasKit renderer), we can just attach that to the renderer object as a single field.

Reply:
We actually have that. See the `CrossFrameCache` class below. Any given type of object can choose which strategy works best. Some could use `CrossFrameCache`. Others would be single-frame scoped. It probably makes sense for paints to be frame-scoped because they can hold onto expensive resources such as images. But for paths maybe cross-frame caching works better?